Source/Core/DolphinQt2/FIFOWindow.cpp

```diff
@@ -0,0 +1,359 @@
+  m_frame_record_count->setValue(3);
+
+  recording_layout->addWidget(new QLabel(tr("Frames to Record:")));
+  recording_layout->addWidget(m_frame_record_count);
+  recording_group->setLayout(recording_layout);
+
+  // Action Buttons
+  m_load = new QPushButton(tr("Load..."));
+  m_save = new QPushButton(tr("Save..."));
+  m_record = new QPushButton(tr("Record"));
+
+  m_save->setEnabled(false);
+  m_record->setEnabled(false);
+
+  m_button_box = new QDialogButtonBox(QDialogButtonBox::Close);
+
+  layout->addWidget(info_group, 0, 0, 1, -1);
+  layout->addWidget(playback_group, 1, 0, 1, -1);
+  layout->addWidget(recording_group, 2, 0, 1, -1);
+  layout->addWidget(m_load, 3, 0);
+  layout->addWidget(m_save, 3, 1);
+  layout->addWidget(m_record, 3, 2);
+  layout->addWidget(m_button_box, 4, 0, 1, -1);
+
+  setLayout(layout);
+}
+
+void FIFOWindow::ConnectWidgets()
+{
+  connect(m_load, &QPushButton::pressed, this, &FIFOWindow::LoadRecording);
+  connect(m_save, &QPushButton::pressed, this, &FIFOWindow::SaveRecording);
+  connect(m_record, &QPushButton::pressed, this, &FIFOWindow::ToggleRecording);
+  connect(m_button_box, &QDialogButtonBox::rejected, this, &FIFOWindow::reject);
+  connect(m_early_memory_updates, &QCheckBox::toggled, this,
+          &FIFOWindow::OnEarlyMemoryUpdatesChanged);
+  connect(m_frame_range_from, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOWindow::OnFrameFromChanged);
+  connect(m_frame_range_to, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOWindow::OnFrameToChanged);
+
+  connect(m_object_range_from, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOWindow::OnObjectFromChanged);
+  connect(m_object_range_to, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
+          &FIFOWindow::OnObjectToChanged);
+}
+
+void FIFOWindow::LoadRecording()
+{
+  QString path = QFileDialog::getOpenFileName(this, tr("Open FIFO log"), QString(),
+                                              tr("Dolphin FIFO Log (*.dff)"));
+
+  if (path.isEmpty())
+    return;
+
+  emit LoadFIFO(path);
+}
+
+void FIFOWindow::SaveRecording()
+{
+  QString path = QFileDialog::getSaveFileName(this, tr("Save FIFO log"), QString(),
+                                              tr("Dolphin FIFO Log (*.dff)"));
+
+  if (path.isEmpty())
+    return;
+
+  FifoDataFile* file = FifoRecorder::GetInstance().GetRecordedFile();
+
+  bool result = file->Save(path.toStdString());
+
+  if (!result)
+    QMessageBox::critical(this, tr("Error"), tr("Failed to save FIFO log."));
+}
+
+void FIFOWindow::ToggleRecording()
+{
+  FifoRecorder& recorder = FifoRecorder::GetInstance();
+
+  if (!m_recording)
+  {
+    // Start recording
+    recorder.StartRecording(m_frame_record_count->value(), [] {
+      QueueOnObject(&Settings::Instance(), [] { Settings::Instance().TriggerFIFORecordingDone(); });
+    });
+
+    m_save->setEnabled(false);
```

Comment:
So when it's recording, the save button is disabled. Should everything else be disabled as well?

Reply:
It would be more consistent I guess (Although it wouldn't have any effect if a user decided to change any of it)...